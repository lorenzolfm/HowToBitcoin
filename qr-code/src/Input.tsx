type InputProps = {
  placeholder?: string,
  value: string,
  onChange: Function,
};

export const Input = ({
  placeholder, value, onChange
}: InputProps) => {
  return (
    <div className="ui input">
      <input
        placeholder={placeholder}
        type="text"
        value={value === '' ? '' : value}
        onChange={(e) => onChange(e.target.value)}
      />
    </div>
  );
}
