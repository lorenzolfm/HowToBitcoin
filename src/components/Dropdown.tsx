import { useState } from 'react';

type Props = {
  children?: (number | string)[];
  items: string[];
}

export const Dropdown = ({ children, items }: Props): JSX.Element => {
  const [open, setOpen] = useState(false);

  const renderedItems = items.map(item => {
    return (
      <div key={item} className="item">{item}</div>
    );
  });

  return (
    <div
      className={`ui dropdown ${open ? 'active visible' : ''}`}
      onClick={() => setOpen(!open)}
    >
      <div className="text">{children}</div>
      <i className="dropdown icon"></i>
      <div
        className={`menu transition ${open ? 'visible' : 'hidden'}`}
      >
        {renderedItems}
      </div>
    </div>
  );
};
