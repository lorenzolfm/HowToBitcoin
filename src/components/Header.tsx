import { Link } from 'react-router-dom';

export const Header = (): JSX.Element => {
  return (
    <div className="ui secondary pointing menu">
      <Link to="/" className="item">Home</Link>
      <Link to="/qrcode" className="item">Qr Codes</Link>
      <Link to="/transactions" className="item">Transactions Overview</Link>
      <Link to="/sha256" className="item">Hash Function Playground</Link>
      <Link to="/mining" className="item">Mining Overview</Link>
    </div>
  );
};
