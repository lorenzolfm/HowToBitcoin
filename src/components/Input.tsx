import 'styles/index.css';

type Props = {
  id: string;
  placeholder?: string;
  value: string;
  onChange: (v: string) => void;
}

export const Input = (
  { id, onChange, placeholder, value }: Props,
): JSX.Element => {
  return (
    <div className="ui labeled input">
      <div className="ui label">{id}</div>
      <input
        id={id}
        placeholder={placeholder}
        value={value}
        onChange={(e) => onChange(e.target.value)}
      />
    </div>
  );
};
