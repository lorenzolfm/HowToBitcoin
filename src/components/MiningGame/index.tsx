import { useEffect, useRef, useState } from 'react';
import { HEIGHT, WIDTH } from './constants';
import { Controller } from './Controller';

let controller: Controller | null = null;

export const MiningGame = () => {
  const [targetSize, setTargetSize] = useState(HEIGHT / 3);
  const [numOfBullets, setNumOfBullets] = useState(1);
  const [attempts, setAttempts] = useState(0);
  const [isGameRunning, setIsGameRunning] = useState(false);

  const ref = useRef<HTMLCanvasElement>(null);

  useEffect(() => {
    controller = new Controller(
      numOfBullets,
      targetSize,
      setAttempts,
      setIsGameRunning,
    );

    if (ref.current)
      controller.init(ref.current);
  }, []);

  const onTargetSizeChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const newSize = parseInt(event.target.value);

    setTargetSize(newSize);
    controller?.setTargetSize(newSize);
  };

  const onNumOfBulletsChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const newNumOfBullets = parseInt(event.target.value);

    setNumOfBullets(newNumOfBullets);
    controller?.setNumOfBullets(newNumOfBullets);
  };

  return (
    <div className="ui segment">
      <div className="ui two column grid">
        <div className="row">
          <div className="ui labeled input column">
            <div className="ui label">Bullets</div>
            <input
              disabled={isGameRunning}
              id="number-of-bullets"
              type="number"
              min="1"
              max="10"
              step={1}
              value={numOfBullets}
              onChange={onNumOfBulletsChange}
            />
          </div>
          <div className="column">Attempts: {attempts}</div>
        </div>
        <div className="row">
          <div className="ui labeled input column">
            <div className="ui label">Target Size</div>
            <input
              disabled={isGameRunning}
              id="target-size"
              type="number"
              min="10"
              max={HEIGHT * 0.5}
              step={10}
              value={targetSize}
              onChange={onTargetSizeChange}
            />
          </div>
          <div className="ui labeled input column">
            <button disabled={isGameRunning} onClick={() => controller?.run()} className="ui primary button column">Go!</button>
          </div>
        </div>
      </div>
      <br />
      <div style={{ textAlign: 'center' }}>
        <canvas
          ref={ref}
          id="canvas"
          width={WIDTH}
          height={HEIGHT}
        />
      </div>
    </div>
  );
};
