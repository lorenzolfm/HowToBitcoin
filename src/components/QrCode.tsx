export const QrCode = (
  { url }: { url: string },
): JSX.Element | null => {
  if (!url) return null;

  return (
    <div className="ui segment">
      <img src={url} alt="qr-code" />
    </div>
  );
};
