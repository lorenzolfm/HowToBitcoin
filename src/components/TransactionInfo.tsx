import { TxInfo } from 'types';
import { Dropdown } from './Dropdown';

type Props = { txData: TxInfo; };

const toBtc = (satoshis: number) => satoshis / (10 ** 8);

export const TransactionInfo = ({ txData }: Props): JSX.Element => {
  const totalInputAmount = txData.vin.reduce(
    (acc, curr) => acc += curr.prevout.value,
    0,
  );

  const totalOutputAmount = totalInputAmount - txData.fee;

  return (
    <div className="ui segment">
      <p>{txData.txId}</p>
      <h2>Transaction details</h2>
      <div className="ui segment grid">
        <Dropdown items={txData.vin.map(vin => `${vin.prevout.scriptpubkey_address} (${vin.prevout.scriptpubkey_type})`)}>
          Used {txData.vin.length} unspent transaction outputs (UTXOs) as input.
        </Dropdown>
        <Dropdown items={txData.vout.map(vout => `${vout.scriptpubkey_address} (${vout.scriptpubkey_type})`)}>
          Send to {txData.vout.length} different outputs.
        </Dropdown>
      </div>
      <div className="ui segment">
        {
          txData.status.confirmed
            ? <p>This transaction was confirmed on block {txData.status.block_height}.</p>
            : <p>This transaction is yet to be confirmed.</p>
        }
        <p>Fee: {txData.fee} satoshis ({toBtc(txData.fee)} BTC).</p>
        <p>Input amount: {totalInputAmount} satoshis ({toBtc(totalInputAmount)} BTC).</p>
        <p>Output amount: {totalOutputAmount} satoshis ({toBtc(totalOutputAmount)} BTC).</p>
      </div>
    </div >
  );
};
