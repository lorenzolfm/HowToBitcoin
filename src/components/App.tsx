import { Routes, Route } from 'react-router-dom';

import { Header } from './Header';
import { QrCodeGenerator } from 'pages/QrCodeGenerator';
import { TransactionsOverview } from 'pages/TransactionsOverview';
import { MiningOverview } from 'pages/MiningOverview';
import { HashFunction } from 'pages/HashFunction';

export const App = (): JSX.Element => {
  return (
    <div className="ui container">
      <Header />
      <Routes>
        <Route path="/qrcode" element={<QrCodeGenerator />} />
        <Route path="/transactions" element={<TransactionsOverview />} />
        <Route path="/sha256" element={<HashFunction />} />
        <Route path="/mining" element={<MiningOverview />} />
      </Routes>
    </div>
  );
};
