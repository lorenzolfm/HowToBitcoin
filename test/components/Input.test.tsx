import { create } from 'react-test-renderer';
import { Input } from 'components/Input';

describe('Input', () => {
  const props = {
    id: 'test',
    onChange: jest.fn(),
    placeholder: 'test',
    value: 'test',
  };

  it('matches snapshot', () => {
    expect(create(<Input {...props} />)).toMatchSnapshot();
  });
});
