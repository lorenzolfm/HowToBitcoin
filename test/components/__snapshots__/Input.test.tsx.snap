// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input matches snapshot 1`] = `
<div
  className="ui labeled input"
>
  <div
    className="ui label"
  >
    test
  </div>
  <input
    id="test"
    onChange={[Function]}
    placeholder="test"
    value="test"
  />
</div>
`;
